File: algo/MinStack.java
import java.util.Stack;

// Approach 1: Stack of Value/ Minimum Pairs;
// Time Complexity: O(1)
// Space Complexity: O(N)
public class MinStack {
    Stack<int[]> s;

    public MinStack() {
        s = new Stack<>();
    }

    public void push(int val) {
        if (!s.isEmpty() && val > s.peek()[1]) {
            s.add(new int[] { val, s.peek()[1] });
        } else {
            s.add(new int[] { val, val });
        }
    }

    public void pop() {
        s.pop();
    }

    public int top() {
        return s.peek()[0];
    }

    public int getMin() {
        return s.peek()[1];
    }
}

// Approach 1: Two Stacks
// Time Complexity: O(1)
// Space Complexity: O(N)
class MinStack2 {

    Stack<Integer> s;
    Stack<Integer> min;

    public MinStack2() {
        s = new Stack<>();
        min = new Stack<>();
    }

    public void push(int val) {
        s.add(val);
        if (min.isEmpty() || val <= min.peek()) {
            min.add(val);
        }
    }

    public void pop() {
        if (s.peek() <= min.peek()) {
            min.pop();
        }
        s.pop();
    }

    public int top() {
        return s.peek();
    }

    public int getMin() {
        return min.peek();
    }
}

File: algo/MinStack.cpp
class MinStack
{
public:
    MinStack()
    {
    }

    void push(int val)
    {
        if (stack.empty() || val < stack.top().second)
        {
            stack.push({val, val});
        }
        else
        {
            stack.push({val, stack.top().second});
        }
    }

    void pop()
    {
        stack.pop();
    }

    int top()
    {
        return stack.top().first;
    }

    int getMin()
    {
        return stack.top().second;
    }

private:
    // val and min;
    stack<pair<int, int> > stack;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */